str0m/src/session.rs

```diff
@@ -626,24 +641,60 @@ impl Session {
 
     fn poll_packet(&mut self, now: Instant) -> Option<DatagramSend> {
         let srtp_tx = self.srtp_tx.as_mut()?;
+        // Only temporary to measure actual queuing delay for BWE work.
+        let real_now = Instant::now();
+
+        let queue_states: Vec<_> = only_m_line(&self.m_lines)
+            .map(|line| line.buffers_tx_queue_state(now))
+            .collect();
 
+        // Figure out which, if any, queue to poll
+        let PollOutcome::PollQueue(queue) = self.pacer.poll_action(now, &queue_states) else {
+            return None;
+        };
```

Comment:
API wise it doesn't look right to double up on the queue states here. 

The pattern everywhere else is:

1. handle_timeout -> decide what to do. Make state change such as deciding which packet goes next.
2. poll_output -> Enact the state change in 1

Philosophically time doesn't move between 1 and 2, so there's no need to redo any calculations based on time in 2.

Reply:
Things within the queue state can and do change between calls to `Pacer::poll_action` that impact the calculations. For example, packets are removed from the queues, which might mean they become empty, this also impacts the total queue size as well as the average queue time for the packets within the queue.

When we end up in the drain queue state the following calling pattern might happen:

1. `Pacer::poll_action` returns `PollOutcome::Nothing` because we need to wait until the next send time
2. We call `Pacer::handle_time` as a side effect of time moving forward
3. We repeatedly call `Pacer::poll_action` until the queues have been drained

Reply:
@k0nserv I think the way to model that would be:

1. Pacer::handle_timeout -> change state
2. Pacer::poll_action -> get poll packet
3. Pacer::poll_timeout -> straight away
4. repeat from 1

I maintain we should not need the queue state update also on `poll_action`.

Reply:
Well, in the instance when we are draining the queues we want to ensure all the packets are sent as a burst i.e. without any time between them and without anything being queued in between.

In libWebRTC they handle this with a loop but of course we cannot loop since we are producing a single packet at the time. What I've done here is used the fact that time only moves forward when we queue packets and handle time to ensure we get the same behaviour. This works like this:


1. When we enter queue draining mode we record the current value of `now`
2. Then any subsequent calls to `poll_action` will calculate the send time for those packets as the instant they were queued. 

This ensures that when we enter the queue draining mode we call `poll_action` until we have drained the queues fully. This assumption breaks down if we allow the time to move forward between these calls

Reply:
Ok. Then I suggest you take ownership of the queue states.

So on `handle_timeout` you do:

`Pacer::get_queue_states(&mut self) -> &mut Vec<QueueState>`

```
let states = pacer.get_queue_states();
states.clear();
loop buffer in buffer_txs {
   states.push(buffer.queue_state());
}
```

And then in `poll_output` we simply work from the queue state that we got in `handle_timeout`.

Reply:
As it currently stands this doesn't work for the following reason.

When we remove a packet from the queue we would need to reduce the `total_queue_time` value based on the amount of time that packet was queued, this information is currently only available within `PacketizingBuffer`. For this to be possible `QueueState` would need to contain information about all the packets in the queue. A similar argument applies to `QueueState::size` which needs to be reduced based on the size of the packet we remove from the queue

Reply:
Ok. Rather than me trying to figure out the solve, can we instead turn it around?

I need `handle_timeout` to do the state change, and `poll_packet` to deliver packets without doing state changes.

How do we achieve that?

Reply:
https://github.com/algesten/str0m/pull/45#discussion_r1121672266

I studied the code a bit more, and I circle back to that this one should be possible. _straight away_ here means now, so time will not move forward just because we do a handle_timeout.

Reply:
It sounds like you are suggesting that every call to `Pacer::poll_action` is followed by a call to `Pacer::handle_time`. This doesn't work at this level, but it would if it applied to `Session::poll_datagram` and `Session::handle_timeout`.

To be clear the problem we end up with here is this: `Session::queue_states` is a snapshot of the packet queues, when we call `Session::poll_datagram` the state of those packet queues can changes causing `Session::queue_states` to no longer be an accurate snapshot.

> I need handle_timeout to do the state change, and poll_packet to deliver packets without doing state changes.

I'm a bit confused about the motivation for this, `Session::poll_packet` already causes state changes(moving the TWCC counter and removing packets from the selected queue). What I'm currently doing is ensuring that the snapshot of the packet queues stays up to date with these state changes that are already happening.